[Webhook] Allow request parsers to return multiple `RemoteEvent`'s

kbond opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Unblocks https://github.com/symfony/symfony/pull/50324
| License       | MIT

I've run into services that send their webhook events in bulk - a single request containing multiple events.

This is one idea to allow this. I'd like to get some input before going further.

TODO:
- [x] Update changelog
- [x] Update/add tests

kbond committed:
[Webhook] Allow request parsers to return multiple `RemoteEvent`'s

alexandre-daubois commented:
Makes sense to me. I think I never came across a service that sends webhook events by batches like this. But to me, that's a nice idea if this exists. I'd be 👍 on this, especially since the implementation is pretty simple!

kbond commented:
> I think I never came across a service that sends webhook events by batches like this.

Until yesterday, I hadn't either. When working on #58252 I discovered Mailtrap does this and then saw https://github.com/symfony/symfony/pull/50324 is blocked because of this.

stof commented:
This is technically a BC break for *callers* of the parser (as the caller needs to be able to deal with an array).

However, as the typical usage of this parser is to be called by the WebhookController of symfony/webhook (while external packages *implement* the interface), I would not oppose shipping that in a minor version of the component, with an explicit mention about that in the changelog and upgrade guide. But I'd like to know what other members of the core team think about it.

stof commented:
Based on the comment in https://github.com/symfony/symfony/pull/50324#issuecomment-2028822948, it seems like Sendgrid also batches webhooks but our parser silently ignore all events except the first one.

kbond commented:
> our parser silently ignore all events except the first one.
😬 

Once multiple messages can be parsed in one request, what are your thoughts on what to do when just one fails?

stof commented:
Each parser can have its own logic to decide when it returns `null` (to ignore the webhook) and when it throws an exception

stof commented on src/Symfony/Component/Webhook/Controller/WebhookController.php:
```diff
@@ -40,12 +40,17 @@ public function handle(string $type, Request $request): Response
+        $events = $parser->parse($request, $this->parsers[$type]['secret']);
 
-        if (!$event = $parser->parse($request, $this->parsers[$type]['secret'])) {
+        if (null === $events) {
```
I think an empty array (i.e. ignoring all events) should be treated as a rejected response too

kbond commented:
I've added tests, updated changelog and upgrade guide. If this is minor BC Break is acceptable, I think this is good to go.

> it seems like Sendgrid also batches webhooks but our parser silently ignore all events except the first one.

When/if merged, I'll fix the sendgrid bridge.

Kocal commented on src/Symfony/Component/Webhook/Controller/WebhookController.php:
```diff
@@ -40,12 +40,17 @@ public function handle(string $type, Request $request): Response
         }
 
-        $this->bus->dispatch(new ConsumeRemoteEventMessage($type, $event));
+        $events = \is_array($events) ? $events : [$events];
+
+        foreach ($events as $event) {
```
Can be optimized to:
```suggestion
        foreach ((array) $events as $event) {
```

kbond replied:
When not an array, this would cast the object to an array, no?

_Edit: I've confirmed this to be the case._

Kocal replied:
Ah, you're right

stof replied:
be careful. For that particular case it works, but if the object can be an object implementing `\Traversable`, it won't be equivalent

OskarStark commented on src/Symfony/Component/Webhook/Client/RequestParserInterface.php:
```diff
@@ -24,11 +24,11 @@ interface RequestParserInterface
      * @throws RejectWebhookException When the payload is rejected (signature issue, parse issue, ...)
      */
-    public function parse(Request $request, #[\SensitiveParameter] string $secret): ?RemoteEvent;
+    public function parse(Request $request, #[\SensitiveParameter] string $secret): RemoteEvent|array|null;
```
We could stay BC with another parameter enabling. This new behavior, couldn't we?

kbond replied:
Can you elaborate a bit on this?

OskarStark replied:
I was thinking about removing the return type and move it to `@return`, then in the method check, if an additional argument was provided like `$returnArrayInsteadOfOneObject` (ofc to verbose) and in this case return `RemoteEvent[] plus raising a deprecation. 🤔

stof replied:
The BC break is the new return type. I don't see how a parameter would help.

OskarStark replied:
```diff
+ /*
+  * @return RemoteEvent|RemoveEvent[]|null
+  */
- public function parse(Request $request, #[\SensitiveParameter] string $secret): RemoteEvent|array|null;
+ public function parse(Request $request, #[\SensitiveParameter] string $secret);
```

Even if we remove the return type?

kbond replied:
Still a BC break for consumers of the interface:

Consider the existing code in someone's app:

```php
if (!$event = $parser->parse($request, $secret)) {
    return ...;
}

$this->bus->dispatch(new ConsumeRemoteEventMessage($type, $event)); // error if $event is an array - this is the BC break
```

fabpot commented:
As we already have 3 bridges that would work correctly without this change, I would consider this BC break as something we need to fix in 7.2.

nicolas-grekas commented on UPGRADE-7.2.md:
```diff
@@ -71,6 +71,14 @@ TwigBridge
+ * [BC BREAK] `RejectWebhookException::parse()` return type changed from
+   `?RemoteEvent` to `RemoteEvent|array<RemoteEvent>|null`. Classes already
+   implementing this interface are unaffected but consumers of this method
+   will need to be updated to handle the new return type.
```
```suggestion
   will need to be updated to handle the new return type
```

nicolas-grekas reviewed (commented):
What about introducing a parseAll method?
`RemoteEvent|array<RemoteEvent>` is tricky to deal with I fear.

stof commented:
Adding a new method will be a BC break for all implementations . And right now, we have dozens of implementation, all in external packages (either in bridge components or third-party) while we have only 1 known caller in the webhook component itself (and so no impacted by a BC break as its update is shipped in the same package).

stof commented on UPGRADE-7.2.md:
```diff
@@ -71,6 +71,14 @@ TwigBridge
+Webhook
+-------
+
+ * [BC BREAK] `RejectWebhookException::parse()` return type changed from
```
```suggestion
 * [BC BREAK] `RequestParserInterface::parse()` return type changed from
```

nicolas-grekas commented:
With an `@method` of course.

nicolas-grekas commented:
Or a new interface.

stof commented:
but that would still force bridges to implement the parse method returning a single one (and so being broken for batched webhooks)

mahono commented:
Just for curiosity: does this somehow allow streaming events to a webhook?

kbond commented:
> Just for curiosity: does this somehow allow streaming events to a webhook?

No, just accounts for some webhook providers that send multiple events in a single request.

kbond commented:
> What about introducing a parseAll method?

I thought about this and then, in `WebhookController`, use `parseAll` when available and `parse` if not (trigger deprecation in this case). I ran into the same thing that @stof describes - a lot more 3rd party code that would need updating.

FYI, I did a search on packagist and couldn't find any open source code that _consumes_ parsers.

stof commented on UPGRADE-7.2.md:
```diff
@@ -71,6 +71,14 @@ TwigBridge
+ * [BC BREAK] `RequestParserInterface::parse()` return type changed from
+   `?RemoteEvent` to `RemoteEvent|array<RemoteEvent>|null`. Classes already
+   implementing this interface are unaffected but consumers of this method
+   will need to be updated to handle the new return type
```
I would also add a sentence saying `Projects relying on the WebhookController of the component are not affected by the BC break.`.

alexandre-daubois commented on UPGRADE-7.2.md:
```diff
@@ -84,6 +84,15 @@ TwigBridge
+   `?RemoteEvent` to `RemoteEvent|array<RemoteEvent>|null`. Classes already
+   implementing this interface are unaffected but consumers of this method
+   will need to be updated to handle the new return type. Projects relying
+   on the WebhookController of the component are not affected by the BC break
```
```suggestion
   on the `WebhookController` of the component are not affected by the BC break
```

fabpot commented:
Thank you @kbond.
